compiler/rustc_passes/src/entry.rs

```diff
@@ -40,30 +41,34 @@ fn entry_fn(tcx: TyCtxt<'_>, (): ()) -> Option<(DefId, EntryFnType)> {
         EntryContext { tcx, attr_main_fn: None, start_fn: None, non_main_fns: Vec::new() };
 
     for id in tcx.hir().items() {
-        let item = tcx.hir().item(id);
-        find_item(item, &mut ctxt);
+        find_item(id, &mut ctxt);
     }
 
     configure_main(tcx, &ctxt)
 }
 
 // Beware, this is duplicated in `librustc_builtin_macros/test_harness.rs`
 // (with `ast::Item`), so make sure to keep them in sync.
-fn entry_point_type(ctxt: &EntryContext<'_>, item: &Item<'_>, at_root: bool) -> EntryPointType {
-    let attrs = ctxt.tcx.hir().attrs(item.hir_id());
+// A small optimization was added so that hir::Item is fetched only when needed.
+// An equivalent optimization was not applied to the duplicated code in test_harness.rs.
+fn entry_point_type(ctxt: &EntryContext<'_>, id: ItemId, at_root: bool) -> EntryPointType {
+    let attrs = ctxt.tcx.hir().attrs(id.hir_id());
     if ctxt.tcx.sess.contains_name(attrs, sym::start) {
         EntryPointType::Start
     } else if ctxt.tcx.sess.contains_name(attrs, sym::rustc_main) {
         EntryPointType::MainAttr
-    } else if item.ident.name == sym::main {
-        if at_root {
-            // This is a top-level function so can be `main`.
-            EntryPointType::MainNamed
+    } else {
+        let item = ctxt.tcx.hir().item(id);
+        if item.ident.name == sym::main {
```

Comment:
Using `item_name` breaks virtually all of the tests. 

I had 

```rust
if ctxt.tcx.item_name(id.def_id.to_def_id()) == sym::main {
...
```
Error for one test:

```
---- [ui] src/test/ui/zero-sized/zero-size-type-destructors.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 101
command: "/home/fausto/Repo/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/home/fausto/Repo/rust/src/test/ui/zero-sized/zero-size-type-destructors.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/home/fausto/Repo/rust/build/x86_64-unknown-linux-gnu/test/ui/zero-sized/zero-size-type-destructors/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/home/fausto/Repo/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/fausto/Repo/rust/build/x86_64-unknown-linux-gnu/test/ui/zero-sized/zero-size-type-destructors/auxiliary"
stdout: none

```

Reply:
Odd. Can you print the value of `item_name` to understand what happens?

Reply:
Looks like it is not returning a name. I hadn't noticed this error before:

> error: internal compiler error: compiler/rustc_middle/src/ty/mod.rs:2028:13: item_name: no name for DefPath { data: [DisambiguatedDefPathData { data: Misc, disambiguator: 0 }], krate: crate0 }

from 
```rust
    pub fn item_name(self, id: DefId) -> Symbol {
        // Look at cross-crate items first to avoid invalidating the incremental cache
        // unless we have to.
        self.opt_item_name(id).unwrap_or_else(|| {
            bug!("item_name: no name for {:?}", self.def_path(id));
        })
    }
```

Reply:
I was a bit weary of the comment on `tcx.item_name()` but I made the change as suggested.

```rust
    /// When possible, this function should be used for cross-crate lookups over
    /// [`opt_item_name`] to avoid invalidating the incremental cache.
```

Reply:
You are right, those comments are wrong and should be updated. https://github.com/rust-lang/rust/pull/96989